core/src/main/java/org/elasticsearch/search/controller/SearchPhaseController.java

```diff
@@ -349,7 +427,8 @@ public InternalSearchResponse merge(ScoreDoc[] sortedDocs, AtomicArray<? extends
         // merge hits
         List<InternalSearchHit> hits = new ArrayList<>();
         if (!fetchResults.isEmpty()) {
-            for (ScoreDoc shardDoc : sortedDocs) {
+            for (int i = sortedDocs.getOffset(0); i < sortedDocs.getOffset(1); i++) {
```

Comment:
also this one is implicit `int i = 0;`? and maybe we can also derive the length from the request? if we can to that, I think we can get rid of the SortedDocs class entirely and just use implicit offsets from the request length / size?